rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,276 @@
+access requests.
+
+Users won’t be asked to think about roles in advance, but will be able to focus
+on finding and requesting access to individual resources.
+
+The existing “role-based” access requests and the new “search-based” access
+requests proposed here are useful for two different scenarios:
+
+- Sometimes it is helpful to request elevated roles in the system, like
+  `db-admin`, to perform some system upgrades, or system wide-troubleshooting.
+  In this case, role-based access requests work great.
+
+- In other cases, it’s helpful to request access to a subset of resources. In
+  this case, search.
+
+### Example Flow
+
+Alice is a member of the group “splunk”.
+
+She gets an alert that “db-1” is malfunctioning.
+
+Alice needs to locate the "db-1" resource and request access to it.
+
+Alice goes to the “Request access” screen and selects: “select access to resources”
+option that is available alongside the “select access to roles” option.
+
+Alice uses standard (fuzzy search) or advanced search (labels) to enter the
+resource name.
+
+She finds a single resource (e.g. "db-1") or a group of resources by that name and clicks
+"request access".
+
+An access request is created. It is sent to the "db-admins" team members via slack,
+email or other configured channel based on the labels on the requested resources
+(we will add label-based routing to access request plugins).
+
+There are two thresholds set in access requests for this group of resources.
+When both Ivan and Mary approve the request, Alice is granted access for one
+hour.
+
+### Query Language and Filter UI
+
+The query language and filter UI is defined by the
+[Pagination and Search RFD](./0055-webui-ss-paginate-filter.md).
+
+### CLI Mock-Up
+
+```bash
+$ tsh request search --kind db --labels 'env=prod'
+Found 2 items:
+
+name kind     id
+db-1 database 388aff7f-459f-4a43-804a-3729854976ab
+db-2 database 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+
+Create access request by:
+> tsh request create --id 388aff7f-459f-4a43-804a-3729854976ab --id 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+```
+
+Users can search by kind, labels, and keywords. The `tsh request search` command
+will output a `tsh request create` command which can be used to request access.
+
+The CLI UX may be improved by allowing users to perform multiple searches,
+"stage" which resources they want to request access to and store this state
+locally, and finally execute the full request without needing each UUID on the
+command line. But this may be too complex and confusing. With this MVP version
+users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
+is fully customizable and scriptable.
+
+### On-Demand SSH
+
+Many times users would not want to search and request access in two steps.
+Teleport will have a new alias flag modifier that will try accessing a node, in
+case if access denied, it will request access for a node all in one step:
+
```

Comment:
Async vs sync mode:

* Early version would not need `async` mode of requests. We might add those in the future if someone askss

How to deal with multiple nodes for `tsh ssh -P root@app=jenkins` for example?

* Look at the implementation complexity and decide whether to add to initial release or postpone. Consider limitations of how many nodes can be added to the request. 

Whether we want to support `-P` for databases and other resources

* Initial release should not focus on anything other than SSH.